plugins/user_mapping/src/local_file.cpp

```diff
@@ -0,0 +1,188 @@
+			[](const auto& _iter) -> user_profile {
+				return {.irods_user_name = _iter.key(), .attributes = _iter.value()};
+			});
+
+		// Move new item list to old one, update last write time
+		std::unique_lock update_profile_list_lock{list_mutex};
+		profile_list = std::move(temp_list);
+		last_file_path_write = std::filesystem::last_write_time(file_path);
+	} // update
+
+	auto match(const nlohmann::json& _params) -> std::optional<std::string>
+	{
+		// If there is an exception while updating, catch so we can still
+		// provide matches with our current good state.
+		try {
+			// Only allow one thread to run update at a time
+			if (std::unique_lock call_update_lock{update_mutex, std::try_to_lock}; call_update_lock) {
+				update();
+			}
+		}
+		catch (const std::exception& e) {
+			spdlog::error("{}: {}", __func__, e.what());
+		}
+
+		std::shared_lock read_profile_list_lock{list_mutex};
+
+		// Use provided mappings to see if there is a complete match to a user
+		for (const auto& [irods_username, attributes] : profile_list) {
+			auto attr_iter{attributes.items()};
+
+			// Verify that each attribute specified is found in _params
+			auto res{std::all_of(std::begin(attr_iter), std::end(attr_iter), [&_params](const auto& _iter) -> bool {
+				const auto value_of_interest{_params.find(_iter.key())};
+				if (value_of_interest == std::end(_params)) {
+					return false;
+				}
+
+				return *value_of_interest == _iter.value();
+			})};
+
+			// If all specified attributes matched, _params maps to irods_username
+			if (res) {
+				return irods_username;
+			}
+		}
+
+		return std::nullopt;
+	} // match
+} // anonymous namespace
+
+auto user_mapper_init(const char* _args) -> int
+{
+	// Check if any of the args are nullptr
+	if (nullptr == _args) {
+		return 1;
+	}
+
+	try {
+		spdlog::debug("{}: Received _args [{}]", __func__, _args);
+		init(nlohmann::json::parse(_args));
+	}
+	catch (const std::exception& e) {
+		spdlog::error("{}: {}", __func__, e.what());
+		return 1;
+	}
+	return 0;
+} // user_mapper_init
+
+auto user_mapper_match(const char* _param, char** _match) -> int
+{
+	// Check if any of the args are nullptr
+	if (nullptr == _param || nullptr == _match) {
+		return 1;
+	}
+
+	spdlog::debug("{}: Attempting match of _param [{}]", __func__, _param);
+	try {
+		auto res{match(nlohmann::json::parse(_param))};
+
+		if (res) {
+			spdlog::debug("{}: Matched _param [{}] to user [{}]", __func__, _param, *res);
+			char* matched_username{strdup(res->c_str())};
+
+			*_match = matched_username;
+			return 0;
+		}
+
+		*_match = nullptr;
+		return 0;
+	}
+	catch (const std::exception& e) {
+		spdlog::error("{}: {}", __func__, e.what());
+		*_match = nullptr;
+		return 1;
+	}
+} // user_mapper_match
+
+auto user_mapper_close() -> int
+{
+	return 0;
+} // user_mapper_close
+
+auto user_mapper_free(char* _data) -> void
+{
+	// NOLINTNEXTLINE(cppcoreguidelines-no-malloc, cppcoreguidelines-owning-memory)
+	std::free(_data);
+} // user_mapper_free
```

Comment:
I'm a little confused by the existence of this function in the interface. Is this just an abstraction for tying up other loose ends in the plugin which may be introduced in the future? Is there a reason `user_mapper_close` wouldn't fill that role?

Reply:
`user_mapper_free(...)` is to be used to free the C-strings from `user_mapper_match(...)`. It's abstracted to allow for other 'free' implementations, based on how the plugin is implemented.

You can see the free function used within `map_json_to_user(...)` here:
https://github.com/irods/irods_client_http_api/blob/f5f4bf3fabba9ba295e44b8c2cd63510115ac828/core/src/common.cpp#L340-L363

I think you can just get away with `user_mapper_close(...)` if you keep track of all allocations made and free them all at the end of the HTTP API execution, but then you have a lot of memory being used (depending on your impl)...

Reply:
Yep!

Reply:
Found the original thread: https://github.com/irods/irods_client_http_api/pull/331#discussion_r1759089287

Reply:
Excellent. Thanks for finding that, and sorry for re-treading things. :) Resolving.